# Ciclos



## 1. La instrucción `for`

Ya hemos visto la instrucción `for`, la cual permite iterar una cantidad determinada de veces. El  esquema general del `for` es
```
for <var> in <iterable>:
    <instrucción(es)>
```
Aquí `<var>` denota una variable cualquiera e `<iterable>` es un objeto iterable,  es decir un objeto que puede ser recorrido secuencialmente, como puede ser una cadena o una lista, un diccionario, un conjunto, etc.,  estructuras que veremos más adelante.

El primer uso que dimos del `for` fue recorrer una lista de enteros de `0` a `n - 1` (para un `n` determinado) y  eso se hace con una instrucción del tipo:
```
for <var> in range(n):
    <instrucción(es)>
```
Es decir `range(n)` es un iterable que contiene todos los números enteros de `0` a `n - 1` en orden. El `for` la va dando a la variable todos los valores de `0` a `n - 1`  y en ese orden. Por ejemplo,


In [ ]:
for i in range(5):
    print(i)

0
1
2
3
4


imprime `0`, `1`, `2`, `3`, `4`, en ese orden.

La  función `range()` además puede tomar otros argumentos: si `m`, `n` son dos enteros,  entonces `range(m, n)` es un iterable  que tiene todos los enteros de `m` a `n - 1`. Por  ejemplo,   

In [ ]:
for i in range(2, 5):
    print(i)

2
3
4


Un iterable conocido son las cadenas. Con el `for` podemos recorrer una cadena desde el primer caracter hasta el último, por ejemplo

In [ ]:
for x in 'diccionario':
    print(x)

d
i
c
c
i
o
n
a
r
i
o


Definamos una función que dada una palabra devuelva la palabra con las vocales eliminadas:

In [ ]:
def eliminar_vocales(cadena: str) -> str:
    # pre: cadena es una cadena de caracteres
    # post: devuelve cadena sin vocales minúsculas
    sin_vocales = ''
    for x in cadena:
        if x != 'a' and x != 'e' and x != 'i' and x != 'o' and x != 'u':
            sin_vocales += x
    return sin_vocales

Usando reglas elementales de lógica (`not (p or q) == not p and not q`) podemos hacer una segunda definición de la función:

In [ ]:
def eliminar_vocales(cadena: str) -> str:
    # pre: cadena es una cadena de caracteres
    # post: devuelve cadena sin vocales minúsculas
    sin_vocales = ''
    for x in cadena:
        if not(x == 'a' or x == 'e' or x == 'i' or x == 'o' or x == 'u'):
            sin_vocales += x
    return sin_vocales

También podemos simplificar un poc la definición de la función con el uso de conjuntos (que veremos más adelante)

In [ ]:
def eliminar_vocales(cadena: str) -> str:
    # pre: cadena es una cadena de caracteres
    # post: devuelve cadena sin vocales minúsculas
    sin_vocales = ''
    for x in cadena:
        if x not in {'a', 'e', 'i', 'o', 'u'}:
            sin_vocales += x
    return sin_vocales

O con listas:

In [ ]:
def eliminar_vocales(cadena: str) -> str:
    # pre: cadena es una cadena de caracteres
    # post: devuelve cadena sin vocales minúsculas
    sin_vocales = ''
    for x in cadena:
        if x not in ['a', 'e', 'i', 'o', 'u']:
            sin_vocales += x
    return sin_vocales

Esta última definición es quizás la más clara.

Probemos la función:

In [ ]:
print(eliminar_vocales('hola'))
print(eliminar_vocales('abracadabra'))
print(eliminar_vocales('murcielago'))
print(eliminar_vocales('Algoritmos y programación')) # elimina vocales en cadenas de caracteres pero no detecta mayúsculas y acentos

hl
brcdbr
mrclg
Algrtms y prgrmcón


### La función `range()`

La función `range()` además de tomar como argumento un entero positivo puede tomar otros argumentos: si `m`, `n` son dos enteros, entonces `range(m, n)` es un iterable que tiene todos los enteros de `m` a `n - 1`. Por ejemplo,

In [ ]:
for i in range(3, 8):
    print(i)

3
4
5
6
7


imprime `3`, `4`, `5`, `6`, `7`, en ese orden. Observar que


In [ ]:
for i in range(8, 3):
    print(i)

Es un conjunto de instrucciones correctas pero no imprime nada porque el iterable `range(8, 3)` es vacío.

La  función `range()` también puede tomar tres argumentos: si `m`, `n`, `k` son tres enteros,  entonces

    range(m, n, k)
    
es un iterable  que tiene todos los enteros desde `m` al entero *anterior* a `n ` "saltando"  de `k` en `k`. Por  ejemplo,   

In [ ]:
for i in range(8, 16, 2):
    print(i)

8
10
12
14


imprime todo los enteros pares de `8` a `15`. Otro ejemplo:

In [ ]:
for i in range(16, 8, -2):
    print(i)

16
14
12
10


imprime todo los enteros pares de `16` a `10` en orden descendente. Es decir,  en este caso  `range(16, 8, -2)` representa los enteros `16`, `14`, `12`,  `10`, en ese orden.   

Es importante observar que `range(m, n, 1)` es lo mismo que `range(m, n)` y  que `range(0, n, 1)` es lo mismo que `range(n)`. Si los argumentos de `range()` no son enteros obtenemos un error.

Cerremos esta sección con un ejemplo:


In [ ]:
def cuenta_regresiva(n: int):
    # pre: n  es un entero positivo
    # post: imprime los enteros de n a 1 y finalmente la palabra ¡Despegue!
    for i in range(n, 0, -1):
        print(i)
    print('¡Despegue!')

cuenta_regresiva(10)

10
9
8
7
6
5
4
3
2
1
¡Despegue!


Observar que fue necesario poner como segundo argumento de `range()` el  número `0` pues el `for` recorre desde el  primer número (en este caso `10`) hasta el anterior a `0`,  que en este caso es `1`, pues el recorrido se hace de mayor a menor (eso es lo que indica el tercer parámetro de `range()`).

## 2. La instrucción `while`

Las computadoras se utilizan a menudo para automatizar tareas repetitivas. Repetir tareas idénticas o similares sin cometer errores es algo que las computadoras hacen bien y las personas hacen mal. En un programa de computadora, la repetición también se llama *iteración*.

Python proporciona instrucciones para hacer iteración. Una es la instrucción `for` que vimos más arriba.

Otra es la intrucción `while`. El uso del `while` nos permite ejecutar una sección de código repetidas veces, de ahí su nombre. El código se ejecutará mientras una condición determinada se cumpla. Cuando se deje de cumplir, se saldrá del bucle y se continuará la ejecución normal. Llamaremos *iteración* a una ejecución completa del bloque de código.

La sintaxis es:
```
while <condición>:
    <instrucción(es)>
```
Mientras se cumpla la `<condición>` se ejecutará el cuerpo del `while`,  es decir las `<instrucción(es)>`, y cuando estas se terminen de ejecutar se  vuelve a comprobar la `<condiciòn>`. Cuando deje de cumplirse la `<condición>`, la ejecución abandonará el `while` y pasará a la instrucción siguiente.

Aquí hay una versión de `cuenta_regresiva()` que usa  `while`:

In [ ]:
def cuenta_regresiva(n: int):
    assert type(n) == int and n > 0, 'el parámetro debe ser un entero positivo'
    """
    pre: n  es un entero positivo
    post: imprime los enteros de n a 1 y finalmente la palabra ¡Despegue!
    """
    while n > 0:
        print(n)
        n = n - 1
    print ('¡Despegue!')

cuenta_regresiva(10)

10
9
8
7
6
5
4
3
2
1
¡Despegue!


Si traducimos del inglés la palabra `while` ("mientras" en castellano) la declaración dice casi textualmente lo que hace: "Mientras `n` es mayor  que `0`, imprimir `n` y luego disminuir `n` en `1`. Cuando llegués a `0`, imprimir `¡Despegue!`"

Más formalmente, aquí está el flujo de ejecución para una declaración `while`:

1. Determinar si la condición es verdadera o falsa.
2. Si es falsa, salir de la instrucción `while` y continuar la ejecución de la siguiente instrucción.
3. Si la condición es verdadera, ejecutar el cuerpo y volver al paso 1.

Este tipo de flujo se llama *bucle* o *ciclo* porque el tercer paso vuelve a la parte superior.

El cuerpo del ciclo debe cambiar el valor de una o más variables para que la condición en algún momento se vuelva falsa y el ciclo termine. De lo contrario, el ciclo se repetirá para siempre, lo que se denomina *ciclo infinito*.

En el caso de `cuenta_regresiva()` podemos probar que el ciclo termina: si `n` es negativo, el ciclo nunca se ejecuta. De lo contrario, `n` se vuelve más pequeño cada vez que pasa por el ciclo, por lo que siempre llegaremos a 0.

Para algunos otros bucles, no es tan fácil de decir. Por ejemplo:

In [ ]:
def collatz(m: int): # tambien llamado "Algoritmo de Siracusa"
    """
    pre: m > 0
    post: imprime el todos los n (intermedios y final) que resultan de aplicar
          el algoritmo de Collatz
    """
    n = m
    i = 0
    while n != 1:
        i += 1
        print(n)
        if n % 2 == 0: # n es par
            n = n // 2
        else: # n es impar
            n = n * 3 + 1
    print(n)
    print('fin de Collatz para m =', m, 'en', i, 'pasos')

collatz(645646456437465346536576756747456746774663655677)


La condición para este ciclo es `n != 1`, por lo que el ciclo continuará hasta que `n` sea `1`, lo que hace que la condición sea falsa.

Cada vez que pasa por el ciclo, el programa genera el valor de `n` y luego verifica si es par o impar. Si es par, `n` se divide por 2. Si es impar, el valor de `n` se reemplaza con `n * 3 + 1`. Por ejemplo, si el argumento pasado a `collatz()` es 3, los valores resultantes de `n` son 3, 10, 5, 16, 8, 4, 2, 1.


In [ ]:
collatz(3)
collatz(31242341) # probar para números grandes

Dado que `n` a veces aumenta y a veces disminuye, no hay pruebas obvias de que `n` alguna vez llegue a 1, o de que el programa finalice. Para algunos valores particulares de `n`, podemos probar la terminación. Por ejemplo, si el valor inicial es una potencia de dos, `n` será par y  mas chico cada vez que pase por el ciclo hasta que llegue a 1. El ejemplo anterior termina con una secuencia de este tipo, comenzando con 16.

Otro ejemplo:

In [ ]:
collatz(2**20)

1048576
524288
262144
131072
65536
32768
16384
8192
4096
2048
1024
512
256
128
64
32
16
8
4
2
1
fin de Collatz para m = 1048576 en 20 pasos


La pregunta difícil es si podemos probar que este programa termina para *todos* los valores positivos de `n`. Hasta ahora, nadie ha podido probarlo o refutarlo, pero se conjetura que el algoritmo siempre termina. Para más información ver en Wikipedia ["Conjetura de Collatz"](https://es.wikipedia.org/wiki/Conjetura_de_Collatz) o el vídeo ["El problema matemático más simple que nadie puede resolver - Conjetura de Collatz"](https://youtu.be/094y1Z2wpJg).

Veamos otro ejemplo del ciclo `while`. Definamos la función `raiz_entera()` que dado un entero no negativo  `n` devuelve la raíz cuadrada entera de `n`.

*Definición.* Sea $n \in \mathbb N$. Entonces $k$ es la *raíz entera de $n$* si $$k^2 \le n < (k+1)^2.$$

Imprimiendo la lista de cuadrados de 1 a 19 podemos ver que la raíz entera de `200` es `14`.

In [ ]:
for i in range(20):
  print(i, i**2)

0 0
1 1
2 4
3 9
4 16
5 25
6 36
7 49
8 64
9 81
10 100
11 121
12 144
13 169
14 196
15 225
16 256
17 289
18 324
19 361


Obviamente el método anterior, "por inspección"  no es el más conveniente para encontrar la raíz entera de un número.

A  continuación, una forma de implemetar la función raíz entera de `n`:

In [ ]:
def raiz_entera(n: int) -> int:
    """
    pre: n >= 0
    post: devuelve k tal que k**2 <= n < (k + 1)**2
    """
    k = 0
    while k**2 <= n:
        k = k + 1
    return k - 1

print(raiz_entera(17))
print(raiz_entera(200))
print(raiz_entera(2000000))

4
14
1414


Observar que el ciclo termina debido a que en cada paso `k` aumenta en 1 y entonces en algún paso `k**2` va a superar a `n`.

En  el ejemplo anterior no podemos reemplazar el `while` por un `for` debido que *a priori* no sabemos cuantos pasos debemos hacer.

**Observación.** Todo ciclo `for` con argumentos enteros puede ser reemplazado fácilmente   por un `while`,  por ejemplo

```
for i in range(n):
    <bloque de código>
```

puede ser reemplazado por

```
i = 0
while i < n:
    <bloque de código>
    i += 1
```

En el caso de trabajar con iterables es un poco más complicado pero es posible hacerlo. No lo veremos en este curso.

## 3. ¿`break` o no `break`?

La respuesta corta es (casi) nunca usar `break`, pero veamos primero que significa esta instrucción.

A veces no es clara cual es la condición para terminar un bucle hasta que se llega a una parte intermedia del cuerpo de instrucciones. En ese caso, es posible utilizar la instrucción `break` para terminar el ciclo.

Por ejemplo, suponga que desea recibir información que el usuario ingresa por teclado hasta que escribe `listo`. Podrías escribir:


In [ ]:
while True:
    linea = input('> ')
    if linea == 'listo':
        break
    print(linea)
print('¡Listo!')

> hola
hola
> adios
adios
> listo
¡Listo!




La condición del bucle es `True`, que siempre es verdadera, por lo que el bucle se ejecuta hasta que llega a la sentencia `break`.

Cada vez ingresa al cuerpo del `while` el programa  le muestra al usuario un corchete angular. Si el usuario escribe `listo`, la instrucción `break` termina el ciclo. De lo contrario, el programa imprime lo que escribe el usuario y vuelve al principio del ciclo. Aquí hay una muestra de ejecución:
```
> Hola
Hola
> listo
¡Listo!
```

Sin embargo, no se recomienda escribir bucles con `break`. En  general, debe controlarse la iteración en el bucle por la condición  del `while`. Agregar un medio para detener repentinamente el ciclo fuera del enfoque normal puede hacer que el código sea difícil de entender y depurar. En el código anterior podría haberse evitado fácilmente el uso del `break` con lo siguiente:

In [ ]:
linea = ''
while linea != 'listo':
    linea = input('> ')
    print(linea)
print('¡Listo!')

> hola
hola
> adios
adios
> listo
listo
¡Listo!


Otro ejemplo de lo que *no* hay que hacer:

In [ ]:
def raiz_entera(n: int) -> int:
    # pre: n >= 0
    # post: devuelve k tal que k**2 <= n < (k + 1)**2
    k = 0
    for x in range(n):
        if x**2 > n:
            k = x -1
            break
    return k

print(raiz_entera(200))

14


La forma correcta de calcular la raíz entera es como se vió en la sección anterior, es decir:

In [ ]:
def raiz_entera(n: int) -> int:
    # pre: n >= 0
    # post: devuelve k tal que k**2 <= n < (k + 1)**2
    k = 0
    while k**2 <= n:
        k = k + 1
    return k - 1

## 4. Ejemplos de uso correcto de ciclos

Como ya vimos (casi) nunca debemos usar `break` en los ciclos.


Veamos otro ejemplo, ahora con el `for`: supongamos  que queremos averiguar si un número `n` es primo o no. Una forma sencilla de hacerlo es ir dividiendo por todos los números de `2` hasta `n-1` y si encontramos que uno de esos número divide a `n` el número es compuesto, en caso contrario es primo. Una implementación rápida podría ser

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    primo = True
    for i in range(2, n):
        if n % i == 0:
            primo = False
            print(i, 'es divisor de', n)
    return primo

print(es_primo(1070567)) # True
print(es_primo(10705671)) # False
# print(es_primo(10705670567056811)) # False (tarda mucho)
# print(es_primo(563456345634569)) # True (tarda mucho)

True
3 es divisor de 10705671
9 es divisor de 10705671
107 es divisor de 10705671
321 es divisor de 10705671
963 es divisor de 10705671
11117 es divisor de 10705671
33351 es divisor de 10705671
100053 es divisor de 10705671
1189519 es divisor de 10705671
3568557 es divisor de 10705671
False
19 es divisor de 10705670567056811


KeyboardInterrupt: 

Observar que esta función verifica *todos* los números menores que el número de input. En particular `es_primo(n)` imprime todos los divisores de `n`. Por lo tanto,  no es una función eficiente, ni mucho menos, especialmente cuando el número es divisible por primos "pequeños".

Con eso en mente, podemos implementar un  código que refleje la siguiente idea: "todo número es primo hasta que se demuestre lo contrario y en ese caso devuelvo `False`".

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    primo = True
    for i in range(2, n):
        if n % i == 0:
            primo = False
            # print(i, 'es divisor de', n)
            break
    return primo

print(es_primo(1070567)) # True
print(es_primo(10705671)) # False
print(es_primo(10705670567056811)) # False
# print(es_primo(563456345634569)) # True (tarda mucho)

True
False
False




Esta implementación es correcta y traduce eficientemente el algoritmo que habíamos pensado.  Sin embargo, "romper" un ciclo `for` con un `break` no es una práctica recomendada. En este caso, donde el código es muy sencillo, no habría problemas, pero para códigos más complejos,  como ya dijimos, puede traer problemas.

Una propuesta sin el uso de `break` puede ser la siguiente:

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    primo = True
    i = 2
    while i < n and primo == True:
        if n % i == 0:
            primo = False
        i += 1
    return primo

print(es_primo(1070567)) # True
print(es_primo(10705671)) # False
print(es_primo(10705670567056811)) # False
# print(es_primo(563456345634569)) # True (tarda mucho)

True
False
False


El código anterior es igual de eficiente que el del `for` con un `break` pero está mejor estructurado.

Observemos que el `for` debió ser reemplazado por el `while` puesto que estamos en el caso de un ciclo que no sabemos cuantas repeticiones tendrá.

Hay muchas formas  de hacer este programa, veamos otra forma. Usualmente podemos usar la negación de la condición que disparaba el `break` como una condición del `while`:

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    i = 2
    while i < n and n % i != 0:
        i += 1
    if i < n:
        return False
    else:
        return True

print(es_primo(563456345634567)) # False

3
False


La función definida arriba tiene dos `return` y eso es algo que debería evitarse, lo ideal es que haya un solo `return` al final del cuerpo de la función.  

Es preferible la siguiente definición:

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    primo = True
    i = 2
    while i < n and n % i != 0:
        i += 1
    if i < n:
        primo =  False
    return primo

print(es_primo(563456345634567)) # False

3
False


O la forma más concisa,  que pareciera ser la mejor con este algoritmo en mente:

In [ ]:
def es_primo(n: int) -> bool:
    # pre: n > 0
    # post: devuelve True si n  es primo, en caso contrario devuelve False
    i = 2
    while i < n and n % i != 0:
        i += 1
    return i == n

## 5. Método de Newton para la raíz cuadrada

El *método de Newton* o de *Newton-Raphson* es un algoritmo para encontrar aproximaciones de los ceros o raíces de una función real. Una explicación completa se puede encontrar en Wikipedia en el artículo [Método de Newton](https://es.wikipedia.org/wiki/M%C3%A9todo_de_Newton).

Se puede utilizar el  método de Newton para encontrar una aproximación numérica de la raíz cuadrada de un número real. Observar que si $a >0$
$$
b = \sqrt{a}\quad \Leftrightarrow \quad b^2 = a \quad\Leftrightarrow\quad b^2-a=0.
$$
Es decir, $\sqrt{a}$ es raíz de la función $f(x) = x^2 -a$. El algoritmo que se obtiene de aplicar el  método de Newton a $f$ se puede describir de la siguiente forma:
1. Sea $x_0$ un valor inicial arbitrario ($>0$).
2. Sea
\begin{equation}
x_{n+1} = \frac12(x_n + \frac{a}{x_n}) \tag{*}
\end{equation}
para $n \ge 0$.

3. Aplique la fórmula (*) desde $n=0$ en adelante hasta obtener un $x_n$ que se aproxime "lo suficiente".

En el ítem 3. la frase "lo suficiente" no es del todo clara y  trataremos de interpretarla a lo largo de la sección.

Veamos como funciona en algún caso particular, por ejemplo, si $a$ es $4$ y $x_0$ es $3$. De ahora en más, por ser más cómodo; llamaremos $x$  a $x_n$ (el valor original) e $y$ a $x_{n+1}$ (el valor que se obtiene). Luego la fórmula (*) se reescribe:
\begin{equation}
y = \frac12(x + \frac{a}{x}) \tag{*}
\end{equation}

Con esta fórmula en mente, la primera aproximación a $\sqrt{4} =2$ es


In [ ]:
a = 4
x = 3
y = (1/2)*(x + a/x)
print(y)

2.1666666666666665


El resultado es `2.1666666666666665` que está más cerca que `3` de la respuesta correcta ($\sqrt{4} = 2$). Si nosotros repetimos el proceso con la nueva estimación, se acerca aún más:


In [ ]:
x = y # el nuevo x es el y  anterior
y = (1/2)*(x + a/x)
print(y)

2.0064102564102564


De la celda anterior obtenemos `2.0064102564102564`. Después de algunas actualizaciones más, la estimación es muy precisa:

In [ ]:
x = y
y = (1/2)*(x + a/x)
print(y)
x = y
y = (1/2)*(x + a/x)
print(y)

2.0000102400262145
2.0000000000262146


Finalmente,  con una aproximación más obtenemos el resultado exacto:

In [ ]:
x = y
y = (1/2)*(x + a/x)
print(y)

2.0


En general, no sabemos de antemano cuántos pasos se necesitan para llegar a la respuesta correcta, pero sabemos cuándo llegamos porque la estimación
deja de cambiar:

In [ ]:
x = y
y = (1/2)*(x + a/x)
print(y)

2.0


Cuando `y == x`, podemos detenernos. Aquí hay un bucle que comienza
con una estimación inicial, `x`, y la mejora hasta que deja de cambiar. Podemos escribir la función `raiz_cuadrada()` de la siguiente forma:


In [ ]:
def raiz_cuadrada(a: float) -> float:
    # pre: a > 0
    # post: devuelve y una aproximación de la raíz cuadrada de a
    x = a / 2 # aproximamos la raíz de a por a/2
    y = (1/2)*(x + a/x)
    while y != x:
        x = y # el último calculado pasa a ser el que se usa para calcular el próximo
        y = (1/2)*(x + a/x)
    return y

print(raiz_cuadrada(4), raiz_cuadrada(4)**2)
print(raiz_cuadrada(9), raiz_cuadrada(9)**2)
print(raiz_cuadrada(30), raiz_cuadrada(30)**2)
print(raiz_cuadrada(101), raiz_cuadrada(101)**2)
print(raiz_cuadrada(102), raiz_cuadrada(102)**2)

2.0 4.0
3.0 9.0
5.477225575051661 30.0
10.04987562112089 101.0
10.099504938362077 101.99999999999999


Para muchos de los valores de `a` esto funciona bien, pero en general no es recomendable probar la igualdad de dos `float`. Los valores de coma flotante son solo aproximadamente correctos: la mayoría de los números racionales, como $ 1/3 $, y los números irracionales, como $\sqrt{2}$, no se pueden representar exactamente con un `float`.

En lugar de comprobar si `x` e `y` son exactamente iguales, es más seguro usar la función `abs` que devuelve el valor absoluto de un número real:  detenemos el bucle cuando `abs(x - y)` es menor que un error que consideramos aceptable. Redefinamos la función:

In [ ]:
def raiz_cuadrada(a: float) -> float:
    # pre: a > 0
    # post: devuelve y una aproximación de la raíz cuadrada de a
    x = a / 2 # aproximamos la raíz de a por a/2
    y = (1/2)*(x + a/x)
    error = 0.0001
    while abs(x - y) > error:
        x = y
        y = (1/2)*(x + a/x)
    return y

print(raiz_cuadrada(4), raiz_cuadrada(4)**2)
print(raiz_cuadrada(9), raiz_cuadrada(9)**2)
print(raiz_cuadrada(30), raiz_cuadrada(30)**2)
print(raiz_cuadrada(101), raiz_cuadrada(101)**2)
print(raiz_cuadrada(102), raiz_cuadrada(102)**2)

2.0 4.0
3.0000000000393214 9.000000000235929
5.477225575302472 30.0000000027475
10.049875621244148 101.00000000247745
10.099504938503184 102.0000000028502


In [ ]:
def raiz_cuadrada(a: float) -> float:
    """
    pre: a > 0
    post: devuelve y una aproximación de la raíz cuadrada de a
    """
    x = a / 2 # aproximamos la raíz de a por a/2
    y = (1/2)*(x + a/x)
    error = 0.00001
    while abs(x**2 - a) > error:
        x = y
        y = (1/2)*(x + a/x)
    return y
n = 4
print('raíz cuadrada de:', n, 'es', raiz_cuadrada(n),'que elevado al cuadrado  es', raiz_cuadrada(n)**2)
n = 9
print('raíz cuadrada de:', n, 'es', raiz_cuadrada(n),'que elevado al cuadrado  es', raiz_cuadrada(n)**2)
n = 30
print('raíz cuadrada de:', n, 'es', raiz_cuadrada(n),'que elevado al cuadrado  es', raiz_cuadrada(n)**2)
n = 101
print('raíz cuadrada de:', n, 'es', raiz_cuadrada(n),'que elevado al cuadrado  es', raiz_cuadrada(n)**2)
n = 102
print('raíz cuadrada de:', n, 'es', raiz_cuadrada(n),'que elevado al cuadrado  es', raiz_cuadrada(n)**2)



raíz cuadrada de: 4 es 2.0 que elevado al cuadrado  es 4.0
raíz cuadrada de: 9 es 3.0 que elevado al cuadrado  es 9.0
raíz cuadrada de: 30 es 5.477225575051661 que elevado al cuadrado  es 30.0
raíz cuadrada de: 101 es 10.04987562112089 que elevado al cuadrado  es 101.0
raíz cuadrada de: 102 es 10.099504938362077 que elevado al cuadrado  es 101.99999999999999


Esta última sería la forma más correcta de escribir la función. La primera forma, sin el error, depende mucho de la implementación del compilador y se pueden obtener resultados inciertos.

## 6. Programación defensiva

El primer paso para obtener las resultados correctos de nuestros programas es asumir que los errores sucederán y protegerse contra ellos. Esto se llama programación defensiva, y la forma más común de hacerlo es agregar aserciones a nuestro código para que se compruebe a sí mismo a medida que se ejecuta. Una *aserción* es simplemente una declaración de que algo debe ser cierto en un cierto punto de un programa. La sintaxis de la aserción es la siguiente

```
assert condición, 'Cadena  que se imprime cuando condición == False'
```



Cuando Python ve una aserción, evalúa la condición de la aserción. Si es cierta, Python no hace nada, pero si es falsa, Python detiene el programa inmediatamente e imprime el mensaje de error explicitado por el programador. Por ejemplo, este fragmento de código suma todos los números de una lista de números positivos:


In [ ]:
numeros = [1.5, 2.3, 0.7, 0.001, 4.4] # lista de números (veremos esto en profundidad más adelante)
total = 0.0
for n in numeros:
    assert n > 0.0, 'Los datos solo deben contener valores positivos (' + str(n) + ' <= 0).'
    total += n
print('El total es:', total)

En  caso que la lista de números contenga un número no positivo,  obtendremos un error por el `assert`:

In [ ]:
numeros = [1.5, 2.3, 0.7, -0.001, 4.4]
total = 0.0
for n in numeros:
    assert n > 0.0, 'Los datos solo deben contener valores positivos (' + str(n) + ' <= 0).'
    total += n
print('El total es:', total)

### Precondiciones y postcondiciones (revisado)
Repasemos el concepto de precondición y postcondición y veamos que en algunos casos pueden verificarse con el comando `assert`. Como ya vimos, las *precondiciones* y *postcondiciones* son anotaciones dentro del cuerpo de una función que nos sirven para saber como deben ser los parámetros que recibe la función y  que es lo que hace la función:

- *Precondición:* es algo que debe ser cierto cuando comienza la la función, para que esta funcione correctamente.
- *Postcondición:* es algo que la función garantiza cierto cuando la misma termina.

Las precondiciones y postcondiciones son parte de la documentación del programa y suelen escribirse como comentarios formales cuando se puede, o en caso contrario informales.

En el caso que se puedan formalizar en Python,  se pueden reemplazar o completar con `assert`.

Veamos un ejemplo con el máximo común divisor. La función escrita en la siguiente celda de código determina  el máximo común divisor de `a` y `b` cuando `a` y `b` son enteros no negativos y uno de ellos es mayor que 0.

El algoritmo se basa en la dos propiedades siguientes: si $x$, $y$  son enteros,  con $y$ no nulo,  entonces:
\begin{align}
\operatorname{mcd}(x, y) &= \operatorname{mcd}(x, y - x) \\
\operatorname{mcd}(0, y) &= y.
\end{align}

In [ ]:
# 1º versión.
def mcd(a, b: int) -> int:
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0"
        x, y = min(x, y - x), max(x, y - x)
    return y

mcd(15, 24)
#mcd(-3, 2)

Esta función se comporta bien mientras `a` y `b` cumplan que son no negativos y al menos uno sea no nulo. Verificar, por ejemplo, que


```
mcd(-3, 2)
```
es una instrucción que no termina.

Agreguemos ahora precondiciones y postcondiciones de manera informal:




In [ ]:
# 2º versión.
def mcd(a, b: int) -> int:
    # pre: a y b no negativos, al menos uno de ellos no nulo.
    # post: devuelve el mcd de a y b
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0"
      x, y = min(x, y - x), max(x, y - x)
    return y

En  la tercera versión la precondición se escribe de manera formal, pero sigue siendo un comentario:

In [ ]:
# 3º versión
def mcd(a, b: int) -> int:
    # pre: (a >= 0 and b >=0) and (a != 0 or b != 0).
    # post: devuelve el mdc de a y b
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0"
        x, y = min(x, y - x), max(x, y - x)
    return y

La tercera versión no difiere en nada, al tiempo de ejecutarse, a la segunda versión, por lo tanto existe la posiblidad de entrar en un ciclo infinito para ciertos valores de `a`  y `b`.

En  la tercera versión hemos formalizado la precondición en lenguaje Python y por lo tanto  la podemos formalizar programáticamente con `assert`:

In [ ]:
# 4º versión
def mcd(a, b: int) -> int:
    # pre: (a >= 0 and b >=0) and (a != 0 or b != 0).
    assert type(a) == int and type(b) == int, "Los parámetros deben ser enteros"
    assert (a >= 0 and b >=0) and (a != 0 or b != 0), "Los parámetros deben ser no negativos, al menos uno de ellos no nulo"
    # post: devuelve el mdc de a y b
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0"
      x, y = min(x, y - x), max(x, y - x)
    return y

print(mcd(15, 24))
# print(mcd(-3, 2))

Esta cuarta versión produce un error y detiene el programa cuando `a` y `b` no cumplen la precondición. En  este caso el `assert` nos garantiza que no se producirá un ciclo infinito y por lo tanto juega un rol importante em el programa.

En el caso de la función `mcd()` formalizar matemáticamente la  postcondición es posible, pero en general la postcondición no se implementa con un `assert`. Por lo tanto la podemos dejar escrita de manera informal o la escribimos con el fomalismo de la matemática, pero sin hacer un `assert` para verificarla.

Por otro  lado, hacer un `assert` para la postcondición no parece razonable, puesto que si la función está bien implementada la postcondición siempre debe cumplirse y si ocurriera en algún caso que no se cumple es porque la función está mal definida.


In [ ]:
# 5º versión
def mcd(a, b: int) -> int:
    # pre: (a >= 0 and b >=0) and (a != 0 or b != 0).
    assert type(a) == int and type(b) == int, "Los parámetros deben ser enteros"
    assert (a >= 0 and b >=0) and (a != 0 or b != 0), "Los parámetros deben ser no negativos, al menos uno de ellos no nulo"
    # post: devuelve d tal que
    #       1) a % d == 0 and b % d == 0, y
    #       2) si c entero y (a % c == 0 and b % c == 0) => d % c == 0
    x, y = min(a, b), max(a, b)
    while x != 0: # "mientras x distinto de 0"
      x, y = min(x, y - x), max(x, y - x)
    return y